source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,210 @@
+class BaseVerifierImpl : public Verifier {
+public:
+  void registerCallback(VerifierCallbacks* callback) override { callback_ = callback; }
+
+  VerifierCallbacks* getCallback(VerifyContext& context) {
+    if (callback_ != nullptr) {
+      return callback_;
+    }
+    return context.callback();
+  }
+
+protected:
+  // The caller's callback
+  VerifierCallbacks* callback_{};
+};
+
+// Provider specific verifier
+class ProviderNameVerifierImpl : public BaseVerifierImpl {
+public:
+  ProviderNameVerifierImpl(const std::vector<std::string>& audiences, const AuthFactory& factory,
+                           const JwtProvider& provider)
+      : audiences_(audiences), auth_factory_(factory),
+        extractor_(
+            Extractor::create(provider.issuer(), provider.from_headers(), provider.from_params())),
+        issuer_(provider.issuer()) {}
+
+  void verify(VerifyContext& context) override {
+    auto auth = auth_factory_.create(audiences_, absl::optional<std::string>{issuer_}, false);
+    auth->sanitizePayloadHeaders(context.headers());
+    auth->verify(context.headers(), extractor_->extract(context.headers()),
+                 [&](const Status& status) { onComplete(status, context); });
+    if (!context.hasResponded(this)) {
+      context.addAuth(std::move(auth));
+    }
+  }
+
+  void onComplete(const Status& status, VerifyContext& context) {
+    if (!context.hasResponded(this)) {
+      context.responded(this);
+      getCallback(context)->onComplete(status, context);
+    }
+  }
+
+private:
+  const std::vector<std::string> audiences_;
+  const AuthFactory& auth_factory_;
+  const ExtractorConstPtr extractor_;
+  const std::string issuer_;
+};
+
+// Allow missing or failed verifier
+class AllowFailedVerifierImpl : public BaseVerifierImpl {
+public:
+  AllowFailedVerifierImpl(const AuthFactory& factory, const Extractor& extractor, bool allow_failed)
+      : auth_factory_(factory), extractor_(extractor), allow_failed_(allow_failed) {}
+
+  void verify(VerifyContext& context) override {
+    auto auth = auth_factory_.create({}, absl::nullopt, allow_failed_);
+    auth->sanitizePayloadHeaders(context.headers());
+    auth->verify(context.headers(), extractor_.extract(context.headers()),
+                 [&](const Status& status) { onComplete(status, context); });
+    if (!context.hasResponded(this)) {
+      context.addAuth(std::move(auth));
+    }
+  }
+
+  void onComplete(const Status&, VerifyContext& context) {
+    if (!context.hasResponded(this)) {
+      context.responded(this);
+      getCallback(context)->onComplete(Status::Ok, context);
+    }
+  }
+
+private:
+  const AuthFactory& auth_factory_;
+  const Extractor& extractor_;
+  const bool allow_failed_;
+};
+
+// Base verifier for requires all or any.
+class BaseGroupVerifierImpl : public BaseVerifierImpl {
+public:
+  void verify(VerifyContext& context) override {
+    context.resetCount(this);
```

Comment:
context is per-request.   AllVerifier() and AnyVerifier may be multiple layers.   2 "all" verifier;  one of them maybe a "any" verifier.   So your counter may have to go with each GroupVerifier in the context.

Reply:
ResetCount is a bad name. I will rename it. This is actually initializing the count for each GroupVerifier to zero.

Reply:
I guess this is not even needed. I'll remove it.